### PR TITLE
Make it possible to partition full model queries

### PR DESCRIPTION
**Pull request by ypc-faros**
## Description

So that we can split reading a full graph into N buckets concurrently.

## Type of change
- [ ] Bug fix
- [X] New feature
- [ ] Breaking change

## Related issues

> Fix [#1]() 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

File: sources/faros-graphql-source/resources/spec.json [hunk 1]
```diff
@@ -84,16 +84,19 @@
         "description": "Legacy V1 schema as a gzip compressed, base 64 encoded text. Only used when 'Adapt V1 query' is enabled.",
         "multiline": true
       },
-      "replace_origin_map": {
```
Comment (ypc-faros) [inline sources/faros-graphql-source/resources/spec.json:87]: I had forgotten to remove `replace_origin_map`. It was moved to the destination in a previous PR

File: sources/faros-graphql-source/src/index.ts [hunk 2]
```diff
@@ -81,6 +84,24 @@ export class FarosGraphSource extends AirbyteSourceBase<GraphQLConfig> {
             " when 'Adapt V1 query' is enabled"
         );
     }
+
+    if (config.query) {
+      if (config.bucket_id !== undefined)
+        throw new VError('Bucket id cannot be used in combination with query');
+      if (config.bucket_total !== undefined)
+        throw new VError(
+          'Bucket total cannot be used in combination with query'
```
Comment (tovbinm) [inline sources/faros-graphql-source/src/index.ts:93]: Let's add this to description

File: sources/faros-graphql-source/src/index.ts [hunk 3]
```diff
@@ -81,6 +84,24 @@ export class FarosGraphSource extends AirbyteSourceBase<GraphQLConfig> {
             " when 'Adapt V1 query' is enabled"
         );
     }
+
+    if (config.query) {
+      if (config.bucket_id !== undefined)
+        throw new VError('Bucket id cannot be used in combination with query');
```
Comment (tovbinm) [inline sources/faros-graphql-source/src/index.ts:90]: Let's also add this to description

**Review by tovbinm — APPROVED** [review]
@cjwooo can you please have a look? This is quite similar to what you did with GitHub source, but now applied to source slices.

**Comment by tovbinm** [comment]
@ypc-faros why not create a slice per model?

**Comment by ypc-faros** [comment]
> @ypc-faros why not create a slice per model?

We do actually. Now that I think about it this probably doesn't help much if Airbyte runs the slices in parallel. Does it @cjwooo ?

**Comment by ypc-faros** [comment]
> @cjwooo can you please have a look? This is quite similar to what you did with GitHub source, but now applied to source slices.

That's exactly where I got inspiration from :)

**Comment by cjwooo** [comment]
Airbyte (or rather, the Airbyte Connector CDK), does NOT run slices in parallel.

**Comment by tovbinm** [comment]
1. What's the benefit having multiple slices by stream?
2. Should we introduce support for parallelism?

**Comment by cjwooo** [comment]
1. Stream slices allow incremental stream state to be updated in a more granular way than the normal interval-based checkpointing, which is only usable if the stream guarantees that records are outputted in ascending order based on the stream's cursor field. Most streams cannot meet that requirement.
2. More investigation is needed. We'd have to see if the Airbyte server will properly handle the output of parallel slices. This could also break our compatibility with community destinations, which may be operating under the assumption that streams and slices are processed sequentially. Performance benefit may also not be worth the effort, given most sources are network bound and often subject to API rate-limiting.

**Review by cjwooo — APPROVED** [review]

**Comment by ypc-faros** [comment]
Thank you for your comments @cjwooo @tovbinm  🙏 !

**Comment by tovbinm** [comment]
@cjwooo @ypc-faros so why offer bucketization instead of just keep slicing by model?

**Comment by ypc-faros** [comment]
> @cjwooo @ypc-faros so why offer bucketization instead of just keep slicing by model?

We can (I'm hoping) speed up copying a graph (by splitting into N concurrent copies). It should also help if there are failures. We can retry copying only the buckets that failed.

**Comment by ypc-faros** [comment]
> > @cjwooo @ypc-faros so why offer bucketization instead of just keep slicing by model?
> 
> We can (I'm hoping) speed up copying a graph (by splitting into N concurrent copies). It should also help if there are failures. We can retry copying only the buckets that failed.

Btw the slicing also stays. We're just providing a way to distribute the slices among N concurrent (or not) syncs.

**Pull request merged**